src/irmin-pack/unix/io.ml

```diff
@@ -108,154 +116,114 @@ module Unix = struct
-          (* really_write and following do not mutate the given buffer, so
-             Bytes.unsafe_of_string is actually safe *)
-          let buf = Bytes.unsafe_of_string s in
-          let () = Util.really_write t.fd off buf 0 (Bytes.length buf) in
-          ()
-    with Unix.Unix_error (e, s1, s2) ->
-      raise (Write_error (`Io_misc (e, s1, s2)))
+    match (t.closed, t.readonly) with
+    | true, _ -> raise (Io_error `Write_on_closed)
+    | _, true -> raise (Io_error `Ro_not_allowed)
+    | _ ->
+        (* really_write and following do not mutate the given buffer, so
+           Bytes.unsafe_of_string is actually safe *)
+        let buf = Bytes.unsafe_of_string s in
+        let () = Util.really_write t.fd off buf 0 (Bytes.length buf) in
+        ()
 
   let fsync t : (unit, [> write_error ]) result =
-    try
-      match (t.closed, t.readonly) with
-      | true, _ -> Error `Write_on_closed
-      | _, true -> Error `Ro_not_allowed
-      | _ ->
+    match (t.closed, t.readonly) with
+    | true, _ -> Error `Write_on_closed
+    | _, true -> Error `Ro_not_allowed
+    | _ -> (
+        try
           Unix.fsync t.fd;
           Ok ()
-    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+        with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2)))
 
   let read_to_string t ~off ~len =
-    try
-      match t.closed with
-      | true -> Error `Read_on_closed
-      | false -> (
-          let buf = Bytes.create len in
+    match t.closed with
+    | true -> Error `Read_on_closed
+    | false -> (
+        let buf = Bytes.create len in
+        try
           let nread = Util.really_read t.fd off len buf in
           match nread = len with
           | false ->
-              (* didn't manage to read the desired amount; in this case the interface seems to
-                 require we return `Read_out_of_bounds FIXME check this, because it is unusual
-                 - the normal API allows return of a short string *)
+              (* didn't manage to read the desired amount *)
               Error `Read_out_of_bounds
           | true ->
-              (* we just created the buf; after we return it as a string, no-one has a handle
-                 on it, so it cannot be mutated; so Bytes.unsafe_to_string is obviously safe
-                 here *)
-              Ok (Bytes.unsafe_to_string buf))
-    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+              (* we just created the buf; after we return it as a string, no-one
+                 has a handle on it, so it cannot be mutated; so
+                 [Bytes.unsafe_to_string] is obviously safe here *)
+              Ok (Bytes.unsafe_to_string buf)
+        with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2)))
 
-  (* NOTE copied from read_to_string above; ensure the code for both functions remains in
-     sync *)
   let read_exn t ~off ~len buf : unit =
-    assert (len <= Bytes.length buf);
-    try
```

Comment:
The API I designed was not right. It did not reflect the fact that we should not open `try` clauses in the `_exn` functions, for performance reasons. The `Unix.Unix_error` exceptions should be propagated up to the irmin user.